### ISSUE TITLE
ceph 18.2.4-3

### ISSUE THREAD

**Issue by bazaah** [bazaah/aur-ceph#26]
Rebuild for fmt 11, reported by [00nave198](https://aur.archlinux.org/account/00nave198) in https://aur.archlinux.org/pkgbase/ceph-bin#comment-99093

**Comment by bazaah** [comment]
ran into instances of https://github.com/fmtlib/fmt/issues/2752 (again) on my hopeful build... will need to put in some work I guess to get this going with fmt 11

**Comment by wlritchi** [comment]
Looks like the [fmt version check](https://github.com/ceph/ceph/blob/reef/cmake/modules/Findfmt.cmake#L3,L21) breaks with fmt 11 because the define was moved to [a new file](https://github.com/fmtlib/fmt/blob/master/include/fmt/base.h#L26) again. Interestingly, Ceph 19 [reworks the version check](https://github.com/ceph/ceph/commit/686dd3d838b6de6bf1101f0b61f729426b0bccad) and falls back to a submodule-built fmt if the version is out of the supported range. A [later PR](https://github.com/ceph/ceph/pull/58990) bumps that submodule up to fmt 11 along with a bunch of changes to support it.

Maybe it's safer/easier to backport the version check/submodule fallback than to backport the fmt changes?

FWIW, I get this build error:
<details><summary>Details</summary>

```
CMake Error at cmake/modules/Findfmt.cmake:15 (math):
  math cannot parse the expression: " % 100": syntax error, unexpected
  exp_MOD (2).
Call Stack (most recent call first):
  src/CMakeLists.txt:327 (find_package)


CMake Error at cmake/modules/Findfmt.cmake:16 (math):
  math cannot parse the expression: "( - ERROR) / 100": syntax error,
  unexpected exp_CLOSEPARENT (10).
Call Stack (most recent call first):
  src/CMakeLists.txt:327 (find_package)


CMake Error at cmake/modules/Findfmt.cmake:15 (math):
  math cannot parse the expression: "ERROR % 100": syntax error, unexpected
  exp_MOD (7).
Call Stack (most recent call first):
  src/CMakeLists.txt:327 (find_package)


CMake Error at cmake/modules/Findfmt.cmake:16 (math):
  math cannot parse the expression: "(ERROR - ERROR) / 100": syntax error,
  unexpected exp_CLOSEPARENT (15).
Call Stack (most recent call first):
  src/CMakeLists.txt:327 (find_package)


CMake Error at cmake/modules/Findfmt.cmake:15 (math):
  math cannot parse the expression: "ERROR % 100": syntax error, unexpected
  exp_MOD (7).
Call Stack (most recent call first):
  src/CMakeLists.txt:327 (find_package)


CMake Error at cmake/modules/Findfmt.cmake:16 (math):
  math cannot parse the expression: "(ERROR - ERROR) / 100": syntax error,
  unexpected exp_CLOSEPARENT (15).
Call Stack (most recent call first):
  src/CMakeLists.txt:327 (find_package)


-- Could not find fmt, will build it
```
</details> 

which seems to suggest that it intends to use the submodule anyway, but I think the error causes the configure step to fail.

**Comment by bazaah** [comment]
I have a successful build for the fmt11 changes, but `$dayjob` comes first; in general I don't backport changes between major ceph versions excluding what the upstream backports itself.

Generally speaking there seems to have been a _lot_ that changed vis-a-vis cmake for 18=>19:

```
git diff --stat v18.2.4..main -- cmake/modules
 cmake/modules/AddCephTest.cmake                   |   36 +-
 cmake/modules/BuildArrow.cmake                    |    4 +
 cmake/modules/BuildBoost.cmake                    |   28 +-
 cmake/modules/BuildFIO.cmake                      |    3 +-
 cmake/modules/BuildOpentelemetry.cmake            |    1 +
 cmake/modules/BuildQAT.cmake                      |   44 +
 cmake/modules/BuildQATzip.cmake                   |   47 +
 cmake/modules/BuildRocksDB.cmake                  |    1 +
 cmake/modules/BuildUBBD.cmake                     |   22 +
 cmake/modules/Builduadk.cmake                     |   53 +
 cmake/modules/Builduring.cmake                    |    4 +-
 cmake/modules/CTags.cmake                         |    9 +-
 cmake/modules/Distutils.cmake                     |    6 +-
 cmake/modules/FindBoost.cmake                     |   38 +-
 cmake/modules/FindLMDB.cmake                      |   61 +
 cmake/modules/FindPython/Support.cmake            | 4475 +++++++++++++++++----
 cmake/modules/FindPython3.cmake                   |  432 +-
 cmake/modules/FindQAT.cmake                       |   40 +
 cmake/modules/FindQATzip.cmake                    |   33 +
 cmake/modules/FindQatDrv.cmake                    |    8 +
 cmake/modules/FindSanitizers.cmake                |   11 +-
 cmake/modules/Findcap.cmake                       |   35 +
 cmake/modules/Findfmt.cmake                       |   61 -
 cmake/modules/Findlibnbd.cmake                    |   33 +
 cmake/modules/Findqatzip.cmake                    |   24 -
 cmake/modules/Finduring.cmake                     |    2 +-
 cmake/modules/Findzfs.cmake                       |   28 -
 cmake/modules/LimitJobs.cmake                     |   44 +-
 cmake/modules/SIMDExt.cmake                       |    5 +
 cmake/modules/code_tests/Sanitizers_fiber_test.cc |   11 +
 30 files changed, 4580 insertions(+), 1019 deletions(-)
```

**Comment by bazaah** [comment]
closed in eebda5b7764c44147969adc586ec7ea1000558ac
